Support keys and configs in SplitClientForTest

GreyTeardrop opened the pull request:
`SplitClientForTest` can be useful for unit-testing code that depends on `SplitClient`, especially in dynamic scenarios where a static config per test class won't work. As of now, `SplitClientForTest` doesn't support customization of split keys or configs - it always returns the same treatment for all keys of a feature flag and always returns null as a split config.

This change extends `SplitClientForTest`, so it's possible to set treatments for each feature flag and key, and also possible to set split config in addition to the treatment name.

GreyTeardrop committed:
Support keys and configs in SplitClientForTest

SplitClientForTest can be useful for unit-testing code that depends on
SplitClient, especially in dynamic scenarios where a static config per
test class won't work. As of now, SplitClientForTest doesn't support
customization of split keys or configs - it always returns the same
treatment for all keys of a feature flag and always returns null as a
split config.

This change extends SplitClientForTest, so it's possible to set
treatments for each feature flag and key, and also possible to set split
config in addition to the treatment name.

GreyTeardrop commented:
@chillaq is this a change you might be interested in reviewing?

nmayorsplit committed:
Merge pull request #621 from splitio/development

Final version

GreyTeardrop commented:
@nmayorsplit any chance you could take a look at this change?

NicoZelaya commented:
Hello @GreyTeardrop ! I'm Nico from Harness FME and I work with the SDK team. 

First of all I'd like to apologize for the lack of response from our end. Then also thank you for your contribution! I agree this would be useful for unit testing and a good addition to this submodule.
I wanted to let you know that I have assigned this one to myself and will be re-distributing it next week, we'll reach out through our CS team if we have any questions. 

Most likely will ask the team to make a few changes like avoiding the type change on the return for .tests() method for backwards compatibility and once we have it in good shape, we'll most likely merge it in! 
Once we have a better assessment by end of next week (or earlier) we'll reach out to let you know what to expect.

Once more, thank you for your contribution and feedback! 

Best regards,
Nico Zelaya.

GreyTeardrop committed:
Merge branch 'master' into test-client-key-result

GreyTeardrop committed:
Fix SplitClientForTest compatibility

Fix compatibility for the SplitClientForTest.tests() method.
Keep it with the old signature, expose new complete mappings via
testMappings() method.

GreyTeardrop commented:
Thank you, @NicoZelaya! I've adjusted the code a bit to keep it as compatible as possible. There should no longer be breaking changes in method signatures. There's technically still a potentially breaking change, as previously `SplitClientForTest.tests()` exposed a mutable internal representation, but now it can only expose a copy that doesn't affect its internal state.
